rpc: simpler setban and new ban manipulation commands

dhruv opened the pull request:
Initially motivated by: https://github.com/bitcoin/bitcoin/pull/19219/files#r442410269

### Objectives:
1. Consolidate `BanMan` functions
2. Throw fewer RPC errors. If user intent is accomplished, the command is successful even if no change was made.

### Changes:
- Remove `BanMan::Ban(CNetAddr&)` and use `BanMan::Ban(CSubNet&)` instead (ips are subnets of one)
  - RPC change: `setban subnet/ip add` now allows adding a ban entry for an ip in an already banned subnet. No error will be thrown so long as the user intent is accomplished.
- Remove `BanMan::Unban(CNetAddr&)` and use `BanMan::Unban(CSubNet&)` instead (ips are subnets of one)
  - RPC change: Unbanning a non-banned subnet will no longer raise an rpc error as user intent is accomplished.
- Remove unused `BanMan::IsBanned(CSubNet&)`
- RPC change: Add `setban ip removeall` to remove all ban entries that include an IP address
- RPC change: Add `listbanned ip` to return all ban entries that include an IP address

gmaxwell commented:
Because bans have specific durations I'm not sure if this is a good change.  Should a 1 hour ban of foo/16 stop you from adding a 1 year ban of foo/24? 

Right now I already have to undo every ban before adding it to make sure it's not rejected and as a result ends up expiring prematurely.  But if any wider banned blocked every narrower ban, I'm not sure how I could programmatically handle that.  I'd have to have crazy logic interrogated the banlist and broke every larger ban into smaller ones when a smaller one with a different time period was added.

dhruv commented:
> I'm not sure how I could programmatically handle that. I'd have to have crazy logic interrogated the banlist and broke every larger ban into smaller ones when a smaller one with a different time period was added.

Totally fair. I can't see either how clients will be able to make tooling with sane logic. I am going to think this through and come back with a better solution. Thanks for the feedback and a potential solution, @gmaxwell !

jnewbery commented:
Concept ACK on making this consistent. I'll review once you have an updated proposal that addresses gmaxwell's input.

dhruv commented:
### Potential solution

Let `(s0, t0)` be an existing ban, where `s0` is a subnet and `t0` is the ban expiration time. The user is now trying to add another ban `(s1, t1)`.

**Case s1 ⊂ s0:**
If t1 <= t0: raise error, already banned. (We could also return success because the end goal is accomplished - it would minimize client error handling)
If t1 > t0: insert `(s1, t1)` into `BanMan`

**Case s1 ⊇ s0:**
If t1 <= t0: insert `(s1, t1)` into `BanMan`
If t1 > t0: remove `(s0, t0)`, add `(s1, t1)` into `BanMan`

**Case no overlap:**
insert `(s1, t1)` into `BanMan`

Thoughts, @gmaxwell, @jnewbery ?

sipa commented:
I think that approach; it's equivalent to treating every ban as a per-IP thing, and remembering the longest-duration ban for each.

One question I have is how to make clearbanned consistent with that? This would be confusing:
* ban `1.2.3.0/24` for 1h
* ban `1.2.0.0/16` for 2h -> overwrites the previous ban
* unban `1.2.3.0/24` -> failure because no such ban exists

I think the most reasonable thing requiring an exact match for *some* ban record when unbanning, but also make it delete all (longer) subranges?

gmaxwell commented:
I don't really like the banning add removal doing fancy logic.  This will make it extremely hard to have external software add bans, even with the newly proposed behaviour.  It's also unlike access-lists on any networking device I've ever worked with.

What is the downside of the obvious behaviour? There is a list of bans, some of which are shorter-than-maximum-CIDR,  add adds to the list, clear removes from it.  If you have overlapping bans, ... so?  You might expect them to have different durations, or they may come from different sources.

The only limitation I'm aware with that approach is that if you want to remove a ban affecting a specific host or subnet more care might be required to remove *all*...  but that could be addressed with a "removeall"  e.g.

```
setban 1.2/16 add
setban 1.2.3/24 add
setban 1.2.3.5/32 remove #does nothing
setban 1.2.3.5/32 removeall  #removes the 1.2/16 and the 1.2.3/24 ban
```

Then you can get exact management of the ban list via add/remove,  but also get a "just make this host work" removal if that's what you want.

dhruv commented:
I agree with the idea that we should interpret `setban ip/mask add time` as "add a ban entry for this address range", and not as "add a ban entry on each single address in the range". This way of thinking allows for overlapping bans.

I also agree that unintended side effects like `setban 127.0.0.1/32 remove` removing the ban entry for `127.0.0.1/24` is not good as it means client scripts need to maintain banlist state and execute logic prior to, and after unbanning.

Further agree, on not returning an error when the user intent is accomplished (by a pre-existing ban entry) even if it does not create a new ban entry - I will update the code to reflect that.

`removeall` if a good idea, perhaps better named `force-remove` and implemented in a follow-up PR? I am open to adding it in here if we believe we need it for atomic commits.

The one thing I am unsure about is not consolidating ban entries where possible. In my rookie opinion, `BanEntry{127.0.0.1/32 200}` should be replaced with `{BanEntry 127.0.0.1/24 400}` instead of just having an additional "dead" entry. Unnecessary ban entries make every call to `BanMan::IsBanned` a tiny bit more expensive, which makes opening new connections, accepting new connections, and `PeerLogicValidation::ProcessMessage/addr` more expensive. I am new here and don't yet have a good intuitive sense for the frequency of those invocations but it seems like an expense to avoid? 

After a couple attempts, I am not able to see how [this proposal](https://github.com/bitcoin/bitcoin/pull/19825#issuecomment-683187822) will mean the banlist will depend on the order of insertions _so long as `banUntil` times are consistent_. I feel sure I'm missing something. Could you please give an example, @gmaxwell ?

gmaxwell commented:
The most common usage is a relative time, so the times won't be consistent. :)

dhruv commented:
Ah, I see. So it's safe to say that the banlist would not depend on the order of insertions.

gmaxwell commented:
I think it would with what you're suggesting.  You're suggesting that if there is a shorter wider block, and I insert a longer narrower block-- both get added.  But if there is a shorter narrower block and I add a longer wider block the shorter block is removed.

If I give both a relative lifetime of 1 day then make the two calls a second apart, then the resulting set of bans depends on the order.

dhruv commented:
@gmaxwell I still think we are in agreement that _equivalent_ `setban` instructions will produce _identical_ banlists. We seem to be disagreeing as to whether the following two sets of RPC instructions are _equivalent_:

Script 1:
```
setban 127.0.0.0/24 add
sleep 1
setban 127.0.0.0/32 add
```

Script 2:
```
setban 127.0.0.0/32 add
sleep 1
setban 127.0.0.0/24 add
```

Unless I am missing something, per the current RPC semantics, these are not equivalent instructions and the node will behave differently. In order for the node to behave differently and comply with the instructions, the banlists will have to be different.

For the use case you are trying to accomplish (atomic, bulk additions with relative timestamps), I think there are two ways we can be successful:
1. Update the scripts to convert relative times into absolute times
2. Let's scope and work on a new, `addall` rpc which can provide bulk atomic commits to the banlist. I volunteer to take it on if we have consensus.

I think another interesting thing this conversation highlights is our expectation of where canonical state lives. i.e. Do we expect that `listbanned` will return a cumulative of past `setban` instructions (thereby agreeing with state maintained client-side), or whether `listbanned` is only responsible for painting a canonical picture of the current ban entries. I personally prefer the latter because: (a) Clients should not assume state cannot change as that assumes there are no other clients and (b) Fewer ban entries seems more efficient for the p2p layer.

dhruv commented:
Current state of this PR:

### What does the current patch at tip `9cc54bd` do?
It tries to make `setban` behave consistently whether it is provided with an IP address, or a subnet (IP address range).

Let `(s0, t0)` be an existing ban, where `s0` is a subnet(a contiguous range of IP addresses) and `t0` is the ban expiration time. The user is now trying to add another ban `(s1, t1)`. If `s1 ⊂ s0 &&  t1 <= t0` the code does nothing because the requested ban is already in place. If `s1 ⊇ s0 && t1 > t0`, we remove `(s0, t0)`, add `(s1, t1)` to consolidate the ban entries. In all other cases we add `(s1, t1)`. More details [here](https://github.com/bitcoin/bitcoin/pull/19825#issue-475008180).

### What is the unaddressed concern expressed by reviewers?
A reviewer feels that we should not consolidate the ban entries, instead allow redundant ban entries in `BanMan::m_banned`. It is their opinion that this will help keep their existing client scripts simpler. The case they make is that the following scripts should generate identical banlists:

Script 1:
```
setban 127.0.0.0/24 add
sleep 1
setban 127.0.0.0/32 add
```

Script 2:
```
setban 127.0.0.0/32 add
sleep 1
setban 127.0.0.0/24 add
```

Note: Per current `setban` RPC semantics, `setban ip/subnet add` means, "ban this ip/subnet for 1 day (86400 seconds)". i.e. `subnet 127.0.0.0/24 add` == `setban 127.0.0.0/24 add 86400`

### What is the author's response?
The author feels that the aforementioned scripts (say, run at t=0) are not equivalent, and therefore produce different banlists. 

Script 1:
```
t=0 "Ban 127.0.0.0/24 until t=86400"
t=1 "Ban 127.0.0.0/32 until t=86401"
```
At `t=1`, the node must remember two ban entries `(127.0.0.0/24, 86400)`, `(127.0.0.0/32, 86401)`.

Script 2:
```
t=0 "Ban 127.0.0.0/32 until t=86400"
t=1 "Ban 127.0.0.0/24 until t=86401"
```
At `t=1`, it suffices for the node to only remember `(127.0.0.0/24, 86401)` as `(127.0.0.0/32 ⊂ 127.0.0.0/24) && (86400 <= 86401)`.

The author also feels that not consolidating ban entries when possible (and cheap) creates expense in looking up `BanMan::IsBanned(subnet)` which is done when new peers are added, or when the `addr` RPC is processed. The author agrees the overhead is seems minimal and infrequent.

### How can reviewers help?
The author is new to Bitcoin development and will continue to keep an open mind. Any case that can be made in interest of the node efficiency or the Bitcoin network, or shows that the proposed change violates established RPC semantics/expectations and therefore is backwards incompatible would help make the case to no longer consolidate ban entries.

jnewbery commented:
I agree with @gmaxwell's suggestion here: https://github.com/bitcoin/bitcoin/pull/19825#issuecomment-683372477 that we shouldn't try to do any clever logic.

I think the behaviour should be:

- for `setban add`:
    - Convert the IP address/net into the canonical CIDR subet representation
    - If that subnet is not already banned, add an entry
    - If that subnet is already banned:
        - if the new expiry is after old expiry, update the entry to the new expiry (i.e. `setban add` can extend an existing ban)
        - if the new expiry is before the old expiry, don't update the entry to the new expiry (i.e. `setban add` can't shorten an existing ban)
- for `setban remove`:
    - Convert the IP address/net into the canonical CIDR subet representation
    - If that subnet is already banned, remove the ban

and then add two new commands:

- `setban list` (only accepts a single IP address, not a subnet): lists all current ban entries that contain that IP address
- `setban removeall` (only accepts a single IP address, not a subnet): removes all current ban entries that contain that IP address

dhruv committed:
[rpc, refactor] Simplify setban logic and consolidate BanMan::Ban() functions

- consolidate Ban() functions
- setban logic is simpler: adding a subset ban will not throw an error

dhruv committed:
[rpc, refactor] Consolidate BanMan::Unban() functions

Unbanning a non-banned subnet should not raise an rpc error

dhruv committed:
[refactor] Remove unused BanMan::IsBanned(CSubNet)

dhruv committed:
[rpc] Add `setban ip removeall` command

dhruv committed:
[rpc] Add `listbanned ip` command

dhruv commented:
@jnewbery @gmaxwell @practicalswift I've changed the code to reflect what we agreed on. 

### Objectives:
1. Consolidate `BanMan` functions
2. Throw fewer RPC errors. If user intent is accomplished, the command is successful even if no change was made.

### Changes:
- Remove `BanMan::Ban(CNetAddr&)` and use `BanMan::Ban(CSubNet&)` instead (ips are subnets of one)
  - RPC change: `setban subnet/ip add` now allows adding a ban entry for an ip in an already banned subnet. No error will be thrown so long as the user intent is accomplished.
- Remove `BanMan::Unban(CNetAddr&)` and use `BanMan::Unban(CSubNet&)` instead (ips are subnets of one)
  - RPC change: Unbanning a non-banned subnet will no longer raise an rpc error as user intent is accomplished.
- Remove unused `BanMan::IsBanned(CSubNet&)`
- RPC change: Add `setban ip removeall` to remove all ban entries that include an IP address
- RPC change: Add `listbanned ip` to return all ban entries that include an IP address

dhruv commented:
Force pushed again to alter commit messages. Ready for review.

glozow commented on src/banman.cpp:
```diff
@@ -159,12 +135,43 @@ bool BanMan::Unban(const CSubNet& sub_net)
+{
+    {
+        LOCK(m_cs_banned);
+        std::vector<CSubNet> entries_to_delete;
+        for (const auto& it : m_banned) {
+            CSubNet sub_net = it.first;
+            if (sub_net.Match(net_addr)) entries_to_delete.push_back(sub_net);
+        }
+
+        if (!entries_to_delete.empty()) m_is_dirty = true;
+        for (const auto&it : entries_to_delete) {
+            m_banned.erase(it);
+        }
```
Why not erase as you go and use `m_banned.size()` to infer `m_is_dirty`, instead of constructing a whole new vector?

jnewbery replied:
Using erase-remove_if is also more efficient - you'll only need to pass over `m_banned` once instead of up to `entries_to_delete.size()` times.

glozow replied:
yeah `erase_if` would be the best but it's C++20 😭

dhruv replied:
Since `map.erase(it)` invalidates `it` I was doing a double-pass to avoid incrementing `it`. However, it turns out `erase` will [return the Iterator](https://en.cppreference.com/w/cpp/container/map/erase) following the last removed element. So I've updated the code to use that.

@jnewbery I am confused about this:
> you'll only need to pass over m_banned once instead of up to entries_to_delete.size() times.

I think the code is only passing over `m_banned` once. Am I missing something? I also tried erase-remove_if but it does not seem to work for associative containers. Is there an example I can learn from?

jnewbery replied:
Sorry. I mistakenly thought that `m_banned` was a vector.

glozow commented on src/banman.cpp:
```diff
@@ -159,12 +135,43 @@ bool BanMan::Unban(const CSubNet& sub_net)
     SweepBanned();
     banmap = m_banned; //create a thread safe copy
+
+    if (filterForIp) {
```
same, I wonder if this could just be a `std::remove_if`

dhruv replied:
Done.

glozow commented on src/rpc/net.cpp:
```diff
@@ -702,7 +686,9 @@ static RPCHelpMan listbanned()
                 "\nList all manually banned IPs/Subnets.\n",
-                {},
+                {
+                    {"ip", RPCArg::Type::STR, "all ips", "If IP is provided, only bans which include the IP will be returned."},
```
`RPCArg::Optional::OMITTED_NAMED_ARG`?

dhruv replied:
Thanks for showing me that. Much better.

glozow reviewed (commented):
Concept ACK thinking of this as a cleanup based on [your description](https://github.com/bitcoin/bitcoin/pull/19825#issuecomment-744235473), Ieft some light code review for now.

dhruv commented on src/test/fuzz/banman.cpp:
```diff
@@ -38,11 +38,6 @@ void test_one_input(const std::vector<uint8_t>& buffer)
         BanMan ban_man{banlist_file, nullptr, ConsumeBanTimeOffset(fuzzed_data_provider)};
         while (fuzzed_data_provider.ConsumeBool()) {
             switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 11)) {
```
Yup - should increase test coverage. Done.

dhruv commented on src/banman.h:
```diff
@@ -77,7 +77,7 @@ class BanMan
     // Removes all ban entries that include net_addr
     void UnbanAll(const CNetAddr& net_addr);
-    void GetBanned(banmap_t& banmap);
+    void GetBanned(banmap_t& banmap, std::optional<const CNetAddr> filterForIp = {});
```
#20671 was merged, so I'm now using `std::optional` instead of `boost::optional`. 

`std::optional<T>` is considered ill-formed, where `is_reference_v<T>` and also if `!is_destructible_v<T>`. So we can't use `std::optional<const CNetAddr&>` and can't forward declare `CNetAddr` and `CSubNet` in this file as incomplete types are not destructible.

jnewbery replied:
I think better than adding this `std::optional<CNetAddr>` would be to just overload the `GetBanned()` function, since most of the logic in the function is now specific to the one case where you pass a filterForIp. Those feel like almost two different functional interfaces.

I think better yet would be to do the filtering logic inside the RPC `listbanned()` function. It's not required by any other callers, so I don't think there's a need to add it to banman. If other callers need the same filtering logic at some point in the future, the logic can be moved into banman at that point.

dhruv replied:
That sounds good. Updated code accordingly.

dhruv commented:
Thank you for the reviews, @glozow @jnewbery. Comments addressed. Rebased. Ready for further review.

dhruv commented:
Rebased and addressed https://github.com/bitcoin/bitcoin/pull/19825#discussion_r553899004

MarcoFalke commented on src/test/fuzz/banman.cpp:
```diff
@@ -38,47 +38,34 @@ FUZZ_TARGET_INIT(banman, initialize_banman)
-                    ConsumeBanTimeOffset(fuzzed_data_provider), fuzzed_data_provider.ConsumeBool());
-                break;
-            }
-            case 1: {
```
After you rebase, you may want to apply this patch: https://github.com/bitcoin/bitcoin/issues/20837#issuecomment-760097060

dhruv replied:
Done in 0dbe88ad3894cd76d52ed196fd2eed39de1296d2

fjahr commented:
Concept ACK

jonatack commented:
Concept ACK, needs rebase

jonatack commented on src/banman.cpp:
```diff
@@ -159,6 +135,25 @@ bool BanMan::Unban(const CSubNet& sub_net)
+                it = m_banned.erase(it);
+                m_is_dirty = true;
+            } else {
+                it++;
```
579eaaa prefer prefix iterator over postfix one (see developer-notes.md)

dhruv replied:
Done

jonatack commented on src/rpc/net.cpp:
```diff
@@ -663,7 +663,7 @@ static RPCHelpMan setban()
                 {
                     {"subnet", RPCArg::Type::STR, RPCArg::Optional::NO, "The IP/Subnet (see getpeerinfo for nodes IP) with an optional netmask (default is /32 = single IP)"},
-                    {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "'add' to add an IP/Subnet to the list, 'remove' to remove an IP/Subnet from the list"},
+                    {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "'add' to add an IP/Subnet to the list, 'remove' to remove an IP/Subnet from the list, 'removeall' to remove all bans that include the single ip address"},
```
579eaaa
```suggestion
                    {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "'add' to add an IP/Subnet to the list, 'remove' to remove an IP/Subnet from the list, 'removeall' to remove all bans that include the single IP address"},
```

dhruv replied:
Done

jonatack commented on src/rpc/net.cpp:
```diff
@@ -678,38 +678,32 @@ static RPCHelpMan setban()
 
+    bool isSubnet = (request.params[0].get_str().find('/') != std::string::npos);
+    if (strCommand == "removeall" && isSubnet) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: setban removeall requires single ip address");
```
579eaaa
```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: setban removeall requires a single IP address");
```

dhruv replied:
Done

jonatack commented on src/rpc/net.cpp:
```diff
@@ -718,23 +712,13 @@ static RPCHelpMan setban()
+        if (node.banman->Ban(subNet, banTime, absolute) && node.connman) {
+            node.connman->DisconnectNode(subNet);
         }
+    } else if(strCommand == "remove") {
```
579eaaa
```suggestion
    } else if (strCommand == "remove") {
```

dhruv replied:
Done

jonatack commented on src/rpc/net.cpp:
```diff
@@ -768,7 +754,21 @@ static RPCHelpMan listbanned()
+        CNetAddr netAddr;
+        LookupHost(request.params[0].get_str(), netAddr, false);
+        if (!netAddr.IsValid()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: Invalid ip address");
```
38abc322
```suggestion
            throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: Invalid IP address");
```

dhruv replied:
Done

jonatack commented on src/rpc/net.cpp:
```diff
@@ -768,7 +754,21 @@ static RPCHelpMan listbanned()
+            if (!ban_sub_net.Match(netAddr)) {
+                it = banMap.erase(it);
+            } else {
+                it++;
```
38abc322 prefer prefix iterator
```suggestion
                ++it;
```

dhruv replied:
Done

jonatack commented on src/banman.cpp:
```diff
@@ -129,22 +109,18 @@ void BanMan::Ban(const CSubNet& sub_net, int64_t ban_time_offset, bool since_uni
             m_is_dirty = true;
         } else
-            return;
+            return false; // nothing to do, longer ban already exists
```
877b14aaeb ISTM this behavior ought to be documented/clarified in the setban rpc help

dhruv replied:
Done.

jonatack reviewed (commented):
Concept ACK, need rebase, a few minor comments on cursory first look

dhruv committed:
[fuzz] removing unused fuzz test branches (invalidates banman seeds)

dhruv commented:
Rebased and comments addressed. Ready for review.

dhruv commented:
Failing test fixed. Ready for review.

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -952,7 +952,7 @@ static const CRPCCommand commands[] =
     { "network",            "getnetworkinfo",         &getnetworkinfo,         {} },
     { "network",            "setban",                 &setban,                 {"subnet", "command", "bantime", "absolute"} },
-    { "network",            "listbanned",             &listbanned,             {} },
+    { "network",            "listbanned",             &listbanned,             {"ip"} },
```
Can remove this diff, and then rebase

dhruv replied:
Thanks for making that easier :)

dhruv commented:
Rebased.

dhruv commented:
Rebased. Ready for further review.

MarcoFalke commented:
Needs rebase to drop already merged hunks in the fuzz tests

naumenkogs commented on src/netaddress.cpp:
```diff
@@ -1116,7 +1116,7 @@ CSubNet::CSubNet(const CNetAddr& addr) : CSubNet()
     case NET_IPV4:
     case NET_IPV6:
-        valid = true;
+        valid = addr.IsValid();
```
what is this change exactly about? Why is it relevant to the given PR?

naumenkogs commented:
Okay, a lot of things here:
1. I like the consolidation because it removes code.
2. I don’t see a big difference between a no-op and throwing an error?
3. I think UnbanAll would be better if it affected everything downwards (all bans for subnets of given net), not upwards. For “unban all above-nets of a specific addr”, I don’t see a use-case (maybe you can point out one?). If you really want to connect to a given addr, use white list?
4. Same for listbanned

1440000bytes commented:
> Remove BanMan::Ban(CNetAddr&) and use BanMan::Ban(CSubNet&) instead (ips are subnets of one)
RPC change: setban subnet/ip add now allows adding a ban entry for an ip in an already banned subnet. No error will be thrown so long as the user intent is accomplished.
Remove BanMan::Unban(CNetAddr&) and use BanMan::Unban(CSubNet&) instead (ips are subnets of one)
RPC change: Unbanning a non-banned subnet will no longer raise an rpc error as user intent is accomplished.

I don't understand why are these changes required? The errors right now look correct.

> RPC change: Add setban ip removeall to remove all ban entries that include an IP address
RPC change: Add listbanned ip to return all ban entries that include an IP address

Concept ACK for `setban ip removeall` and `listbanned ip`

achow101 commented:
Are you still working on this?

dhruv commented:
@achow101 yes i'd like to, if there's enough interest in the idea. i'm going to work through getting BIP324 PRs up to date with the new mechanisms introduced in the BIP recently and then I can rebase this.

dhruv commented:
I'm ok with marking this up for grabs or closing it.
